python-lib/tagger.py

```diff
@@ -52,29 +62,87 @@ def _get_patterns(self):
         """instanciates PhraseMatcher with associated tags"""
-        _, _ = self.nlp_dict[language].remove_pipe("sentencizer")
+        self.nlp_dict[language].remove_pipe("sentencizer")
         matcher = PhraseMatcher(self.nlp_dict[language].vocab)
         self.patterns = list(self.nlp_dict[language].tokenizer.pipe(self.patterns))
         matcher.add("PatternList", self.patterns)
         self.matcher_dict[language] = matcher
 
-    def _matching_method_with_category(self, language: AnyStr):
+    def _match_with_category(self, language: AnyStr) -> None:
         """Instanciates EntityRuler with associated tags and categories"""
         self.nlp_dict[language].remove_pipe("sentencizer")
         ruler = self.nlp_dict[language].add_pipe("entity_ruler")
         ruler.add_patterns(self.patterns)
 
-    def _matching_pipeline(self):
+    def _format_with_category(self) -> pd.DataFrame:
+        self._match_with_category(self.language)
+        if self.mode == OutputFormat.ONE_ROW_PER_DOCUMENT_JSON.value:
+            formatter = self.instanciate_class(FormatterByDocumentJson)
+            formatter.tag_columns = ["tag_json_full", "tag_json_categories"]
+
+        if self.mode == OutputFormat.ONE_ROW_PER_DOCUMENT.value:
+            formatter = self.instanciate_class(FormatterByDocument)
+            formatter.tag_columns = ["tag_keywords", "tag_sentences"]
+
+        if self.mode == OutputFormat.ONE_ROW_PER_TAG.value:
+            formatter = self.instanciate_class(FormatterByTag)
+            formatter.tag_columns = [
+                "tag_keyword",
+                "tag_sentence",
+                "tag_category",
+                "tag",
+            ]
+            return formatter.write_df()
+        return formatter.write_df_category()
+
+    def _generate_unique_columns(self, columns):
+        text_df_columns = self.text_df.columns.tolist()
+        return [generate_unique(column, text_df_columns) for column in columns]
+
+    def _format_no_category(self) -> pd.DataFrame:
+        self._match_no_category(self.language)
+        if self.mode == OutputFormat.ONE_ROW_PER_DOCUMENT_JSON.value:
+            formatter = self.instanciate_class(FormatterByDocumentJson)
+            formatter.tag_columns = ["tag_json_full"]
+        if self.mode == OutputFormat.ONE_ROW_PER_DOCUMENT.value:
+            formatter = self.instanciate_class(FormatterByDocument)
+            formatter.tag_columns = self._generate_unique_columns(
+                [
+                    "tag_keywords",
+                    "tag_sentences",
+                    "tag_list",
+                ]
+            )
+        if self.mode == OutputFormat.ONE_ROW_PER_TAG.value:
+            formatter = self.instanciate_class(FormatterByTag)
+            formatter.tag_columns = self._generate_unique_columns(
+                ["tag_keyword", "tag_sentence", "tag"]
+            )
+        return formatter.write_df()
+
+    def instanciate_class(self, formatter):
```

Comment:
It's not really standard to have a method to instantiate another class. Can you deal with this more gracefully with class inheritance?

Reply:
I honestly can't find a way to do so without instanciating each class with all its argument... If you have ideas I'll appreciate them very much ! :)

Reply:
Nico T. to the rescue 🦸